### DISCUSSION TITLE
v9.2 (2021/12)

### DISCUSSION THREAD

**Discussion by Mizux** [google/or-tools#3023] [Announces]
# Platform changes
* Add support for Ubuntu 21:10 (last rolling release).

# Dependencies Update
* .Net TFM update `net5.0` -> `net6.0` (need .Net SDK 6.0 LTS and .Net SDK 3.1 LTS).
* abseil-cpp `20210324.2` -> `20211102.0`.
* Protobuf `3.18.0` -> `3.19.1`.
* Googletest `1.10.0` -> `1.11.0`.
* Python: add `numpy >= 1.13.3`.
* On MacOS compile Coin-OR in `-O1` to avoid crash in runners.

# Routing
* Improvement on filters.
* Improve first solution heuristics.
* Improve time break placements.

# CP-SAT
* Breaking changes:
  *  The underlying protocol buffer is incompatible with previous versions. Any stored protocol buffer will have to be re-generated with the updated builder APIs (in C++, Python, Java, and .NET)
  * In particular, the interval protobuf  was clean as we removed the old fields (start, size, and end) and renamed the new ones (using `_view`) to use the name of the removed fields. 
* New features:
  * The `all_different`, `reservoir`, `modulo`, `multiplication` and `division` constraints accept affine expressions (`a * var + b`) everywhere it required integer variables.
  * The objective accepts floating point coefficients (See the `DoubleLinearExpr` class in C++/Java/.NET. See the `knapsack_2d_sat.py` example in Python).
  * The `no_overlap_2d` constraint supports optional intervals.
  * The C++ API implements `+` and `*` operators to build expressions. 
* Improvements:
  * Improved presolve code.
  * Tighter model checker.
  * Rework reservoir constraint.
  * Add energetic cuts for the no_overlap_2d constraint.
  * Improved linear relaxation of encoding constraints (`literal implies var == value`).
* Deprecated and removed methods
  * Deprecated C++ `BooleanSum` and `BooleanScalProd`. Just use `Sum` and `ScalProd`.
  * Removed C++ `AddLinMinEquality` and `AddLinMaxEquality`. Just use `AddMinEquality` and `AddMaxEquality`.
* Future incompatibilities
  * At some point in the future, we will rewrite the Java modelling layer to be closer to the C++ layer.
  * In the C++ modelling layer, we will make the IntVar(BoolVar var) ctor explicit.
  * We are contemplating making the python API PEP 8 compliant (using snake_case names). If this happen, we will provide a sed file to port the code.

# Build System
## Bazel
* Fix Windows build.
## CMake
* Add `FETCH_PYTHON_DEPS` option (default `ON`).
* Add optional support for GPLK solver (default `-DUSE_GLPK=OFF`).

# Python
* Support `numpy` integers in most of the CP-SAT API.
* Fix missing `__version__`.

<hr /><em>This discussion was created from the release <a href='https://github.com/google/or-tools/releases/tag/v9.2'>v9.2 (2021/12)</a>.</em>

**Comment by hklarner** [comment]
- where can I read details about the routing improvements?
- can you share a link to the respective commits (I tried but could not identify the relevant commits)

**Reply by lperron** [reply]
https://docs.github.com/en/repositories/releasing-projects-on-github/comparing-releases

**Reply by hklarner** [reply]
Sure. So here is the relevant link:
- https://github.com/google/or-tools/compare/v9.1...v9.2

And clicking on "load more commits" until the list is exhausted and searching for "routing" results in two commits:
- [ROUTING] remove old gtl:: obsolete code; fix overflow error: https://github.com/google/or-tools/commit/d54ed8f29e3acdba104a8174dae045d0474b158f
- [ROUTING] major improvement on filters/first solution heuristics/time…: https://github.com/google/or-tools/commit/edb66aaed5500b985ba73aa2ec89f24f350f7f01

**Comment by Woodz** [comment]
Any timeline on publishing this release to Maven-central?

**Reply by Mizux** [reply]
done

**Comment by lperron** [comment]
When maven works again.

Le ven. 17 déc. 2021, 06:58, Richard Wise ***@***.***> a
écrit :

**Comment by Mizux** [comment]
https://search.maven.org/artifact/com.google.ortools/ortools-java
still not displayed on maven central: https://mvnrepository.com/artifact/com.google.ortools/ortools-java `-_-`

**Reply by Woodz** [reply]
Looks like it is now displayed - thanks team!
